enhancements/etcd/protecting-etcd-quorum-during-control-plane-scaling.md

```diff
@@ -0,0 +1,452 @@
+
+#### Story 3
+
+As an end user of OpenShift, I want to be able to increase the size of my Control Plane without having to know the
+intricacies of etcd and protecting its quorum.
+
+### API Extensions
+
+This enhancement does not introduce any new API extensions.
+
+### Implementation Details/Notes/Constraints
+
+#### Requirements for etcd safety
+
+* The number of voting members should equal the desired number of Control Plane Machines (and this should be an odd
+  number)
+  * We will only deviate from this to add a replacement member
+  * Once the new member is added, we should remove the old member as soon as possible to reduce the risk of degrading the
+    cluster while having an even number of voting members
+* Existing voting members should not be removed until their replacements have been promoted to voting members
+  * By starting new etcd members as [learners](https://etcd.io/docs/v3.3/learning/learner/#raft-learner), we can ensure
+    that the new member is fully "caught-up" and promotable to a full voting member before we start the removal process
+    of the old member
+  * This protects etcd from potential inconsistencies in its data if the cluster were to have some interruption shortly
+    after a new member joins the cluster
+  * This also ensures that we keep the full etcd data on disk on a minimum of the desired number of Control Plane
+    Machines at all times
+
+#### Protecting etcd during scaling operations
+
+To ensure that the safety requirements described above are maintained during scaling operations,
+the etcd operator will manage the etcd cluster in clusters with a
+[functional Machine API](#When-is-Machine-API-Functional),
+by leveraging [Machine Deletion Hooks](#What-are-Machine-Deletion-Hooks) to coordinate with the Machine API when it is,
+and isn't safe, to remove Machines with voting members of the etcd cluster running on them.
+
+##### Overview of protecion mechanism
+
+To ensure the safety of the etcd cluster quorum, the etcd operator will leverage a pre-drain [Machine Deletion Hooks](#What-are-Machine-Deletion-Hook) to prevent the removal of any Control Plane Machine hosting a voting member of the etcd quorum.
+
+```yaml
+lifecycleHooks:
+  preDrain:
+  - name: EtcdQuorumOperator
+    owner: clusteroperator/etcd
+```
+
+The etcd operator will apply the hook to a Machine resource once it identifies that the Machine hosts an etcd member.
+The hook should be added before the member is promoted to ensure that there is no period where the member is a voting
+member, while the machine is not protected by the deletion mechanism.
+
+It will only remove the hook once it has identified that the Machine resource is being deleted and a replacement member
+has been created. The removal of this hook will allow the Machine API to drain and terminate the Machine as it would
+normally do.
+
+In the case that a Machine is deleted before the member is promoted, the etcd operator is expected to not promote the
```

Comment:
It sounds like we would need some sort of locking to ensure a new member is not promoted when a machine was deleted.

Reply:
I have just realized this bit refers to deleting a new learner member.

Reply:
also, what happens when a newly promoted machine gets deleted? Do we have to guarantee it won't be replaced on the next run?

Reply:
I don't think we actually need to do anything special here. 
The deletion hook should already be on the member before it is promoted.

If we promote a member but the Machine was deleted, all this does is create extra toil.
When we next reconcile it will just look like the Machine was deleted, the promotion element doesn't actually effect this so we just treat it as a voting member being removed. 

This is very much an optimisation rather than a necessity and the design of the hooks and the particular points at which they are added and removed ensure we should be safe.